src/model/implementation/object-type-base.ts

```diff
@@ -41,40 +66,59 @@ export abstract class ObjectTypeBase extends TypeBase {
         }
 
         this.validateDuplicateFields(context);
+        this.validateSystemFieldOverrides(context);
 
         for (const field of this.fields) {
             field.validate(context);
         }
     }
 
+    private validateSystemFieldOverrides(context: ValidationContext): void {
+        for (const systemFieldOverride of this.systemFieldOverrides.values()) {
+            const systemField = this.getSystemFieldOrThrow(systemFieldOverride.name);
+            if (systemField.type.name !== systemFieldOverride.typeName) {
+                context.addMessage(
+                    new ValidationMessage(
+                        Severity.Error,
+                        `System field "${systemField.name}" must be of type "${systemField.type.name}"`,
+                        systemField.astNode,
+                    ),
+                );
+            }
+
+            // id validation on a RootEntity is done in the validate method of RootEntityType
+            if (
+                (systemFieldOverride.name !== 'id' || this.isChildEntityType) &&
+                !systemFieldOverride.isHiddenASTNode
+            ) {
+                context.addMessage(
```

Comment:
an idea to remove the tight coupling: add a method `isSystemFieldOverrideRedundant()`, fill it with the `isHidden` logic here, and add additional logic for `id` in the `RootEntityType`?

also: we usually use the astNode fields only for reporting (so they stay optional for e.g. a programmatic call of the model constructors), actual logic checks are done using the "regular" properties like `isHidden` here

Reply:
That is a good idea 👍 

So I should rely on `isHidden` being `undefined`? I found it easier to reason that when the ASTNode is not there, that the directive was not applied.

Reply:
Didn't think about `false` vs. `undefined`, I would have just said it's non-redundant if `isHidden` is `true`

but I understand your thought - you're saying the field is not redundant as long as the directive is applied

it's not affecting behavior but just a (non-critical) validation, so it's probably fine either way